oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -184,18 +184,44 @@ Imposing Constraints on Dependencies
 Although we usually want to use the latest available version of our
 dependencies in order to take advantage of the latest bug fixes, performance
 improvements, and security fixes, we sometimes need to impose some constraints
-on the version to be used.  These should be collected in
-``requirements/constraints.txt`` so they can be imposed uniformly across all
-the repository's requirements files; this is done via a ``-c constraints.txt``
-line just under the summary comment of each ``*.in`` file in the
-``requirements`` directory.  Some guidelines to keep in mind when populating
-this file:
+on the version to be used. These constraints can either be placed directly on
+the existing dependency in the ``*.in`` file or as a new entry in
+``requirements/constraints.txt``:
+
+* If it is a constraint on a dependency listed in **a library's base.in**,
+  the constraint should be added to the existing dependency line.
+
+  * This allows constraints to be advertised via ``setup.py`` and taken
+    into account in dependency resolution of relying packages.
+  * If a package advertises an extra feature, constraints should be similarly
+    placed in the extra's corresponding ``*.in`` file.
+
+* All other constraints should instead be added to
+  ``requirements/constraints.txt``.
+
+  * The ``*.in`` files must include the line ``-c constraints.txt`` just below
+    their summary comment. This allows constraints to be shared uniformly
+    across multiple files. (The ``-c`` flag keeps the constraints from being
+    interpreted as dependencies.)
+  * Most constraints could be listed either here or in ``*.in`` files directly,
+    but the constraints file was historically preferred and so continuing to
+    prefer it reduces churn. The only case where the constraints file is
+    specifically required is when multiple ``*.in`` files declare the same
+    dependency, but in such a situation it is generally preferred to rearrange
+    things so that one ``*.in`` pulls in the other's ``*.txt`` file  using a
+    ``-r`` line, and only list the dependency in the latter.
```

Comment:
I think I would be more clear that in general we want maximum constraints in the constraints file, unless it must be in a file like base.in.  I think this makes it simpler to go to a single location to understand the list of constraints and remove them as needed. I think this bullet adds additional complexity, to an already complex list of bullets, and doesn't provide enough benefit.

Maybe the doc could get an updated with either my reasoning of why we want to keep these together, or, if there is a reason why we don't want to, help provide that background instead.

Reply:
I actually don't have a strong preference for where "other" constraints should go. (Other: Singularly-specified, and one or more of non-minimum, non-base, or IDA.) I think that most of our max constraints actually will be in base.in, which removes a lot of the usefulness of constraints.txt as a centralized place to find constraints. But it's also easy to collect all the constraints with a single shell command, even if they're spread out among multiple `.in` files: `grep -Po '^[^#;]+[>=]=[^# ]+' requirements/*.in`

It might actually be simpler to switch to preferring any new constraints to be in-place, and say that constraints.txt is _only_ for new constraints that need to be referenced from multiple places.

Reply:
Would we be better off just making the `load_requirements` function in `setup.py` smarter, so that it looks in `constraints.txt` for an override of each package in the `*.in` file being processed?  This would avoid adding a bunch of new logic to the OEP guidelines for people to follow, and keep all our version restrictions in one easily scanned place.  If possible, I'd rather have the computers do extra work instead of developers.

Reply:
I like @jmbowman's proposal.

Reply:
I think I'd like the idea of `load_requirements` being smarter if we only had one repo, but as it is we end up with different versions of that logic across over 60 repos. That logic would also be nontrivial—you'd have to determine if each constraint is applicable to base.in or not by comparing it to base.txt, and possibly also deal with URL-based dependencies. It would be nice to package that up, but there's a bootstrapping problem.

Alternatively, `make upgrade` could perform the extraction and dump it into a place that load_requirements can simply read straight in, and there isn't a bootstrapping problem at that time.

But how much benefit do we actually get from having all the constraints in constraints.txt, versus base.in + constraints.txt?

Reply:
I was thinking of just one addition to the existing logic: before returning the list currently generated by `load_requirements`, see if there's any line in `constraints.txt` that starts with the list entry followed by a character that isn't valid in a package name or URL.  If so, replace the list entry with that line.  The logic shouldn't be that complex, and it could be written to only need to parse `constraints.txt` once.  I can think of one or two pathological corner cases (like when you have different constraints for the same package in `base.in` and `constraints.txt`) where it could produce less than optimal results, but they'd still be better than what we currently have.

And yeah, it's annoying to have to make the logic change in lots of different `setup.py` files.  But I'm concerned that adding complexity on where to put constraints will just result in it being done differently everywhere as developers forget the logic or keep following the old advice.

Reply:
Hmm, I just realized that we could use semgrep or pylint to automate upgrading the setup.py code. Something like this, with the code-cleanup running and `semgrep -f rules.yml --autofix setup.py`:

```yaml
rules:
- id: reqs-constrained
  languages:
  - python
  message: |
    Use a newer load_requirements function that knows how to combine base.in
    with constraints.txt.
  pattern: |
    def load_requirements(...):
      ...
    def is_requirement(line):
      ...
  fix: |
    def load_requirements(constraints_path, *requirements_paths):
        etc
  severity: WARNING
```

Since this is a pretty plausible path forward for replacing code blocks quickly, maybe I'll do an experiment to see how bad it is to implement a smarter load_requirements. (I'd still have some suggested changes for this OEP around min constraints, though.)

Reply:
(Note: Filed ARCHBOM-1772 for tracking.)